### ISSUE TITLE
AAudio full-duplex latency increasing over time reason identified and needs fix to dataCallback() API in AAudio

### ISSUE THREAD

**Issue by stereomatch** [google/oboe#165]
**Summary:  Since Oboe uses AAudio for Oreo 8.1, this is relevant for Oboe as well.  The dataCallback() API should be changed to allow for a frameCount value to be returned to AAudio, or at least a return code which tells AAudio to not add this frame to playback pipeline.**  AAudio should then either buffer non-full buffers, or at least know when buffer is partial (currently AAudio assumes all buffers are full size).  Reason this is relevant is because dataCallback() sometimes gets less than full buffer from recording thread - so it gets a 240 sized buffer most of the time, but then occasionally it gets that as partial buffers (example: 128 and then 1ms second later 112 in a subsequent dataCallback()).  So occasionally two calls to dataCallback() (and thus process()) are generated by AAudio - only dataCallback() knows the size of the buffer actually received (see "frameCount" below) from recording thread, but that info is never communicated with AAudio.  Instead dataCallback() fills the rest of buffer with zeros to make a full 240 sized buffer - and the same for the next partial frame as well - effectively adding a full 240 size buffer worth of data to the playback pipeline.  This means that AAudio is being sent one extra 240 sized buffer occasionally.  This means playback pipeline is being fed nonsense data (which means it will space out the real data) - effectively making latency seem longer and longer over time (until it reaches plateau as playback pipeline starts shedding frames as it's FIFO queue is overflowing).  This matches the observation of latency on Nexus 4 running Oreo 8.1 and OnePlus 5T running Oreo 8.1, where it has been observed that latency seems to increase over time.

**Correction: I removed the logcat from AudioEffect.cpp - process() and instead buffered "numFrames" values to memory, and then flushed it to logcat after 1000 frames (since 240 is a 5ms frame on Nexus 4 running Oreo 8.1, this means after 5 seconds).  Now I did not see any of the partial frames - all numFrames seemed to be 240 i.e. full frames.  We are not supposed to do too much processing in the process() method - however, the stress test did still reveal a weakness in the dataCallback() API - because dataCallback() cannot inform AAudio which frame to ignore, when the occasional partial frame (128 + 112) type situation does occur, it will inject an EXTRA full frame into the playback pipeline.  So the core observation may still be correct - in practice with full duplex apps using AAudio, there WILL occur some time when Android is stressed - and when that occurs, this partial frame situation may occur, and will be pushing an extra frame needlessly into playback pipeline (and there is nothing that dataCallback() can do to prevent it currently).  So a return status code at the very least must be introduced in dataCallback() which tells AAudio to not push this frame into playback pipeline.  dataCallback() then then do aggregation of partial frames itself.**

This is a summarization of the comments ending in the following thread (which is now closed):

https://github.com/google/oboe/issues/80#issuecomment-406123187
how do I query output latency for echo cancellation #80

AAudio echo sample latency has been observed to be changing - usually starts off with minimal latency, and then a little later it seems to have larger latency.

This effects is noticeable by humans if you listen to a click sound - which you hear over the air, and then as output from AAudio echo sample a bit later.

This effect can be reproduced by having an android device with headset listen to a click sound, then have another android device running an audio recorder app (or use a digital audio recorder) listen to the output of the 1st device's earphones.

Now when you click that sound is recorded by the second device, and a little later when that sound emanates from the first device's earphones, it is again heard by the second device.

This way the second recording device is standing in for a human listener.  By taking the recorded audio files and opening in Audacity on your computer, you can measure the time interval between the first click sound that is heard directly over the air, and the second click sound (that emanates from earphones of 1st device).

This latency seems to vary - starting off at minimum latency, and increasing with time.

Latency Test Data

Here is an example of the click separation as viewed in Audacity for the OnePlus 5T running Oreo 8.1 - you can observe there is some variation in latency from 0.34ms to 0.42ms and one with 104ms:

NOTE: each horizontal timeline is one experiment i.e. the impact of a single click and what was heard by 2nd audio device audio recorder app.  Each image below has multiple rows - each row is a separate click.  Thus each row gives an estimate for latency by measuring interval between first two pulses, and as you compare between rows, you can see that the latency seems to be different each time.

- [2018_05_11_oneplus5t_headsetmic_headsetearphone_130ms_32ms_40ms.png](https://imgur.com/a/3ooL2jd)

Here is the same for the Nexus 4 running Oreo 8.1 - using headset mic and headset - giving about 80ms:

- [2018_05_11-01_50_26_nexus4_oreo81_headsetmic_headsetearphones_80ms.png](https://imgur.com/a/BbLQtQt)

But then there were other examples where this varied - from 75ms to 85ms for the Nexus 4:

- [nexus4_oreo81_headsetmic_headsetearphones.png](https://imgur.com/a/5dArgl4)

And examples where it varied much more - for Nexus 4 device mic, and headset earphones - about 110ms:

- [nexus4_oreo81_nexusmic_headsetearphone.png](https://imgur.com/a/pU3exJi)

The OnePlus 5T also varied a bit - for device mic and headset earphones, it varies from 100ms to 155ms:

- [2018_05_11_oneplus5t_devicemic2_150ms.png](https://imgur.com/a/UrHyJtZ)

In AAudio echo sample, let's examine the EchoAudioEngine.cpp - dataCallback(), and the AudioEffect.cpp - process() - here dataCallback() is registered to be called by AAudio - and dataCallback() in turns calls AudioEffect.cpp - process() periodically.

On the Nexus 4 running Oreo 8.1 device that I tested on (the OnePlus 5T running Oreo 8.1 was not available), the process() method is called with 240 frames (for sample rate of 48000Hz it means every 5ms - which seems to match the timestamps on the logcat entries).

While process() is called every 5ms with 240 sized frame, almost every 8th or 9th time, a less than 240 size is set for numFrames in process() - it turns out this is then immediately (within 1 ms) leads to a second call to process with the rest of the buffer (to complete the 240).

It turns out that dataCallback() (which is calling process()) is reading from recording thread (see code snippet below) - and frameCount from that is sometimes less than 240.  The problem however is that in dataCallback() whenever a sub-240 size is delivered (frameCount < numFrames in dataCallback() method) - then the rest of the buffer is filled with zeros.

Since dataCallback() only sends back a status message to AAudio (and doesn't send back frameCount):

	https://developer.android.com/ndk/reference/group/audio.html#group___audio_1gad88ad53a723807879ba75d3b5e07c073
	AAudioStream_dataCallback

	dataCallback() currently returns a status code (and not a frameCount):
	- AAUDIO_CALLBACK_RESULT_CONTINUE
	- AAUDIO_CALLBACK_RESULT_STOP

this means AAudio has no way of knowing - and this means what started as a split frame i.e. instead of 240, we occasionally get 128 and 112 - these wind up creating two complete 240 sized frames.  So basically one extra 240 worth of data is being pushed into the playback pipeline.

This would explain why playback latency seems to increase over time, until it reaches a maximum (probably because of overflow as playback pipeline cannot handle extra buffers and starts dropping a few).

In the AAudio echo demo, in AudioEffects.cpp - process(), if you log the "numFrames" value, these are the types of values you get on a Nexus 4 running Oreo 8.1 (custom ROM):

0
240
240
240
240
240
240
240
240
128
112
240
240
240
240
240
240
240
16
224
...

As can be seen from EchoAudioEngine.cpp - dataCallback(), the frameCount value is local to dataCallback() and is not passed back to AAudio:

	aaudio_result_t frameCount = 0;
	...
	frameCount = AAudioStream_read(recordingStream_, audioData, numFrames, static_cast<int64_t>(0));

      ConvertMonoToStereo(static_cast<int16_t *>(audioData), frameCount);

      audioEffect_.process(static_cast<int16_t *>(audioData), outputChannelCount_, frameCount);
	...

	/**
	* If there's not enough audio data from input stream, fill the rest of buffer with
	* 0 (silence) and continue to loop
	*/
	numFrames -= frameCount;
	if (numFrames > 0) {
		memset(static_cast<int16_t *>(audioData) + frameCount * outputChannelCount_,
		     0, sizeof(int16_t) * numFrames * outputChannelCount_);
	}
    return AAUDIO_CALLBACK_RESULT_CONTINUE;

This means that extra data is being added to the playback pipeline - which inevitably will make recorded data appear later and later in the playback output.  Thus leading to a sense that latency is increasing with time.

**Solution:  Since Oboe uses AAudio for Oreo 8.1, this is relevant for Oboe as well.  The dataCallback() API should be changed to allow for a frameCount value to be returned to AAudio.**  AAudio should then either buffer non-full buffers, or at least know that buffer is empty, so it does nothing (not output anything to playback pipeline).

**Comment by stereomatch** [comment]
Correction: I removed the logcat from AudioEffect.cpp - process() and instead buffered "numFrames" values to memory, and then flushed it to logcat after 1000 frames (since 240 is a 5ms frame on Nexus 4 running Oreo 8.1, this means after 5 seconds).  Now I did not see any of the partial frames - all numFrames seemed to be 240 i.e. full frames.  We are not supposed to do too much processing in the process() method - however, the stress test did still reveal a weakness in the dataCallback() API - because dataCallback() cannot inform AAudio which frame to ignore, when the occasional partial frame (128 + 112) type situation does occur, it will inject an EXTRA full frame into the playback pipeline.  So the core observation may still be correct - in practice with full duplex apps using AAudio, there WILL occur some time when Android is stressed - and when that occurs, this partial frame situation may occur, and will be pushing an extra frame needlessly into playback pipeline (and there is nothing that dataCallback() can do to prevent it currently).  So a return status code at the very least must be introduced in dataCallback() which tells AAudio to not push this frame into playback pipeline.  dataCallback() then then do aggregation of partial frames itself.

**Comment by stereomatch** [comment]
As an easy test to measure latency which anyone can do - plug in headset and keep volume high.

Now make a clicking sound - the feedback from headset earphones to headset mic will be heard as a trrrr.. type of sound initially.

After a few minutes of running the AAudio echo app, try the clicking sound again.  It will now sound like a t.t.t.t... type of sound.

So even audibly you will notice that latency has increased.

This has been observed on Nexus 4 running Oreo 8.1 (custom), and OnePlus 5T running Oreo 8.1 (official), and by some of our users.

**Comment by stereomatch** [comment]
@dturner any chance a change to the API could make it into Android P ?  Also have you noticed this increasing latency over a few minutes with the AAudio echo sample.

**Comment by stereomatch** [comment]
To reproduce the latency variation, run AAudio echo sample - choose:
Recording device: built-in microphone
Playback device: built-in earphone speaker
Keep volume low enough that you don't have heavy feedback, but if you snap your fingers, you should hear a pulse train - 2 or 3 echoes at least.

Now listen to the earphone speaker with your ear, as you snap your fingers.  Each snap will create a pulse train (echoes due to feedback).

Initially the pulse train will be very close together, but then 20 seconds later the pulse train period will noticeably seem to be longer.

However, a bit later the pulse train period will again become smaller.

This suggests the audio latency doesn't just increase monotonically and then become stable.  Instead it increases, then is reset to the minimum, then a bit later it is more.  This suggests there is resetting going on of the playback pipeline (or something like that) - which is resetting the latency back to low levels.  However there is a great deal of variation in it (which means there is probably data loss on playback pipeline as well).

I urge AAudio engineers to do this test on their android devices (Oreo 8.1) to see if they can reproduce this issue.

**Comment by stereomatch** [comment]
Thanks.  That would be great if that was experimentally observed - i.e. if the shortfall in read buffer was one-time (whether by callback or via non-blocking read() method).

However, it has been experimentally observed to be an ongoing/repeating cycle - which repeats in stairstep fashion on:
- Nexus 4 running Oreo 8.1
- OnePlus 5T running Oreo 8.1

as described earlier above:

>Initially the pulse train will be very close together, but then 20 seconds later the pulse train period will noticeably seem to be longer.
>
>However, a bit later the pulse train period will again become smaller.
>
>This suggests the audio latency doesn't just increase monotonically and then become stable. Instead it increases, then is reset to the minimum, then a bit later it is more. This suggests there is resetting going on of the playback pipeline (or something like that) - which is resetting the latency back to low levels. However there is a great deal of variation in it (which means there is probably data loss on playback pipeline as well).


The problem would be manageable if after a second it dies down, and from that moment on the full buffer size is always delivered.

But that is not the case.  It makes sense why it should not be so - i.e. a constant size audio data delivery maybe impractical for a pre-emptively multi-tasked thread that is doing the non-blocking read().  Whenever there is a momentary delay in execution of the read() (i.e. if it is not regular), there will be situations where read() cannot deliver the full data.

This is how it happens with AudioRecord/AudioTrack - where partial audio buffer does get delivered.

However, if the developer is not informed of actual data that non-blocking read() delivers - the developer will then needlessly be pushing out extra data on the output pipeline.  There does not seem to be a rationale why this absence of information is a "good thing".

This is what has been observed - the developer continually pushes a "full" buffer to output - as a result the output pipeline eventually "sheds" some of it due to "overrun" (this is where the monotonically rising latency then suddenly comes back to minimum - as observed above).


To a layman, the simplest solution to this would be that the callback or non-blocking read() method should deliver two pieces of info to the developer:
- the max size of the returned buffer (if this is guaranteed to be a max size, then can be conveyed to developer at the start)
- the actual data in the audio buffer

This simple change would fix the bulk of the problem (in my opinion).


In summary the problem is that the process that happens supposedly only for the first 1-2 seconds, is actually happening on ongoing basis - latency monotonically increases, then suddenly is reset back to minimu latency.  This process repeats over and over.  The absence of feedback of actual data that read() returns is a needless oversight which is not helping anyone.

**Comment by stereomatch** [comment]
Another possibility is that the "shedding" (overrun handling) that AAudio currently does should be done more aggressively.

This way the latency would not get a chance to increase to high levels.  That is, the current behavior of steadily rising latency, then suddenly being reset back to the minimum latency, and then this whole process repeating - that could be mitigated somewhat if that resetting back to minimum latency occurred more frequently.

At some point AAudio is realizing that it is getting way too much data from the developer than it can put out - and so it sheds some of that audio (overrun handling).  That step should perhaps be done more frequently (?)

Anyway, this is just a suggestion - I am not privy to the internal handling in AAudio, so this suggestion maybe missing the mark.

**Comment by stereomatch** [comment]
Phil Burk:

>The best arrangement is to have the output stream use a callback and the input stream use non-blocking read().

Maybe that is the issue - we are using the AAudio echo sample code - where there is a single process() method.

There, even when the developer does not modify the audio buffer, still the sawtooth latency behavior emerges.

Ideally the echo sample code should also be modified to work with the full-duplex sample code that eventually appears.


Is the sample code created by ggfan using a non-blocking read() and separate callback ?

https://github.com/google/oboe/blob/master/samples/LiveEffect/README.md
LiveEffect Sample


One would think that if there was some interest in pushing AAudio or Oboe, it would be higher priority to first push out the demo sample code showing full duplex.

On the basis of what is said on Google I/O, AAudio/Oboe looks interesting, but having demo code would be a better advocate than the spoken word at Google I/O.

**Comment by atneya** [comment]
@philburk can this be resolved by #60

**Comment by philburk** [comment]
I believe this bug is about the the problems inherent in full duplex audio.
A full duplex system is initially in an unstable state where there may not be enough input data available.
That will result in the output callback not having enough data and it will glitch and slip in time relative to the input.
Eventually the input and the output will stabilize and the latency will also be stable.

I was not able to reproduce this issue on any current versions.

LiveEffect now uses the FullDuplex algorithm. It passes through three stages.

1. Drain the input as much as possible.
2. Let the input fill up slightly so it is not running on empty.
3. Run for a while letting the input catch up to output.
4. Stable operation that tries to keep the input buffer empty and the output buffer full.

https://github.com/google/oboe/blob/master/samples/LiveEffect/src/main/cpp/FullDuplexStream.cpp
